aesthetic_ants/penguin/tile_layer.py

```diff
@@ -0,0 +1,113 @@
+        self.tile_images = {
+            TileType.WATER: WATER_TILE,
+            TileType.ICE: ICE_TILE,
+            TileType.WEAK_ICE: WEAK_ICE_TILE,
+            TileType.WALL: WALL_TILE
+        }
+
+        self.x = x
+        self.y = y
+
+        self.batch = Batch()
+
+        # Tiles is in the format [y][x]
+        self.tiles = []
+
+    def set_tile(self, tile_x: int, tile_y: int, tile_type: TileType):
+        """
+        Creates a new tile at a given position in the tile grid
+        """
+        if self.tiles[tile_y][tile_x]:
+            self.tiles[tile_y][tile_x].batch = None
+            self.tiles[tile_y][tile_x].delete()
+
+        self.tiles[tile_y][tile_x] = Tile(tile_type,
+                                          self.tile_images[tile_type],
+                                          x=tile_x * TILE_SIZE + self.x,
+                                          y=tile_y * TILE_SIZE + self.y,
+                                          batch=self.batch)
+
+    def update(self, dt, **kwargs):
+        pass
+
+    def draw(self):
+        self.batch.draw()
+
+    def load_tiles(self, tile_filename):
+        """Loads a tile map from a level file. Levels are 40x30 tiles by default"""
+        with open(tile_filename, 'r') as level_file:
+            level_data = level_file.readlines()
+            self.tiles = []
+
+            for y, line in enumerate(level_data):
+                row = []
+                for x, serialized_tile in enumerate(line.strip()):
+                    tile_type = TILE_SERIALIZATION_MAP[serialized_tile]
+                    tile = Tile(tile_type,
+                                self.tile_images[tile_type],
+                                x=x * TILE_SIZE + self.x,
+                                y=y * TILE_SIZE + self.y,
+                                batch=self.batch)
+                    row.append(tile)
+                self.tiles.append(row)
+
+    def collide_tiles(self, other, _):
+        """Determine which tiles collide with the passed object,
+        and call collide_tile for each"""
+        # Calculate the width and height in tiles of the object
+        tile_width = floor(other.width * other.collision_leniency / TILE_SIZE)
+        tile_height = floor(other.height * other.collision_leniency / TILE_SIZE)
+
+        # Calculates the tile position of the object
+        tile_position = ((other.x - self.x) / TILE_SIZE,
+                         (other.y - self.y) / TILE_SIZE)
+
+        # Use the tile position as a base, making a small list of tile coordinates
+        # that collide with the object
+        tiles = [(floor(x + tile_position[0]), floor(y + tile_position[1]))
+                 for x in range(tile_width)
+                 for y in range(tile_height)]
+
+        # Filter out tiles that have invalid coordinates
+        tiles = filter(lambda pos: len(self.tiles[0]) > pos[0] >= 0
+                       and len(self.tiles) > pos[1] >= 0,
+                       tiles)
```

Comment:
Might be a good idea to store these two `len` calls to avoid calling it multiple times.

Also I prefer either doing this in the for-loop below or using a generator expression, but that's just me.